sql: rework how snowball tracing and EXPLAIN(TRACE) works

andreimatei opened the pull request:
cc @knz

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13221)
<!-- Reviewable:end -->

andreimatei committed:
server, util/tracing: rework the server-side snowball machinery

- introduce RecorderTrace - a canonical "recorder callback" to pass to
  basictracers. Besides clients not having to write recorders, the idea
  is also to provide a good location for the recorded spans to be
  accumulated, in lieu of each client choosing a more or less arbitrary
  location to store them.
- make TeeTracer more flexible - allow a TeeSpan to own or not own each
  of the children spans. Not owning will the useful on the client-side
  (see next commit).
  The spans on the server-side of a snowball RPC are now passed to both
  the regular tracer and a recording tracer; before this patch the
  regular tracer was ignored so, for example, the traces were missing
  from Lightstep. The client side spans will come in the next commit.

andreimatei committed:
sql: switch TRACE_SQL to use the new-style Recorder on the client-side

This commit breaks EXPLAIN(TRACE) which doesn't yet use the new client.

andreimatei committed:
sql: rework how EXPLAIN(TRACE) works

Before, the ctor of the explain node was hijacking the transaction's
context with one that had a snowball trace. Now, I've localized the
hijacking to TraceNode.Next()/Finish(). This is conceptually simpler
and makes the code better (we can get rid of Txn.CollectedSpans), but
has subtle consequences - we're no longer tracing the process of
creating the plan for the EXPLAIN statement (stuff like getting table
leases and resolving some ranges). I think you can argue both ways on
this, and on a related note you can argue whether tracing of the
"preparation" phase for a statement was good or not. In any case, the
plan is to get rid of EXPLAIN(TRACE) all-together in favor of a
mechanism for tracing at the SQL session level.

Also change the EXPLAIN(TRACE) test from an INSERT to a SELECT since
EXPLAIN(TRACE) INSERT/UPDATE/DELETE is not supported (the test was quite
bogus).

Fixes #9564

andreimatei committed:
util/tracing: remove JoinOrNew

This method was a bit too magic and it's not needed any more.

andreimatei commented:
I'm still working on good tests and the exact schema of `session_info.latest_trace`; I might do away with the cumulative duration and compute it at query time.
This series of commits breaks an `EXPLAIN(TRACE)` test because of problems with how MemoryAccounts interact with the session context - discussed in https://reviewable.io/reviews/cockroachdb/cockroach/10478. I want to delete EXPLAIN(TRACE) all-together, but the problem fundamentally remains and I'll work on the fix.

For an example of how a trace looks, see
https://gist.github.com/andreimatei/4a79e99e4894f28b46a3f8901e83924f

RaduBerinde commented:
CC @tamird, he may want to look over this stuff.

Looking at the first commit right now. It turned out to be less than ideal to embed Tracers in contexts; now we are embedding recorders.. any insight as to why this is different/better in principle?

---

Review status: 0 of 27 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/server/node.go, line 896 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbfTcOxVoA713mlUGUR:-KbfTcOxVoA713mlUGUS:byvfp9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/server/node.go#L896)):*
> ```Go
> 		// If this is a "snowball trace", we'll need to encode all the recorded
> 		// spans in the BatchResponse at the end of the request.
> 		if isSnowball {
> ```

I can't see where do we set isSnowball to true? (checked this commit and the entire PR too..)

---

*[pkg/server/node.go, line 898 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbfRtzKHRsAueXY_dnH:-KbfRtzKHRsAueXY_dnI:btvuj2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/server/node.go#L898)):*
> ```Go
> 		if isSnowball {
> 			// Encode all the spans into the BatchResponse.
> 			recorder := tracing.RecorderFromCtx(opCtx)
> ```

It's very unclear in my head where this recorder is coming from. Wouldn't it always be the one set up in `JoinRemoteTrace`? (in which case why can't we use it directly instead of retrieving it through `opCtx`?) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-KbfR7v45CZdXwGbnJsR:b-wvw7d3)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Well the `Tracer` is in the context, right? (the `Span` has a reference to it). Holding an additional, separate reference in the Context was the unfortunate part, I think. The `Recorder` is not directly referenced by the `Tracer` (which can be a `TeeTracer`. So I would put it directly in the `Context`, but uses of it can be limited to the functions in `tracer.go`. In particular, the `IngestRemoteSpans()` function is the important use. 

---

Review status: 0 of 27 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/server/node.go, line 896 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbfTcOxVoA713mlUGUR:-KbfZjvJ_wAt2dQ2Opep:bgfywkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/server/node.go#L896)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I can't see where do we set isSnowball to true? (checked this commit and the entire PR too..)
</blockquote></details>

oops, rebase fail. Fixed.

---

*[pkg/server/node.go, line 898 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbfRtzKHRsAueXY_dnH:-KbfYxNk-mna7eznVcIO:bahiwiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/server/node.go#L898)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's very unclear in my head where this recorder is coming from. Wouldn't it always be the one set up in `JoinRemoteTrace`? (in which case why can't we use it directly instead of retrieving it through `opCtx`?) 
</blockquote></details>

correct. `JoinRemoteTrace` is the only case where a Recorder needs to be used here (there could otherwise be a Recorder active, e.g. in the case of a local "rpc call", but this function would not be responsible for doing anything special with it in that case).
Yes, it could could be returned by `JoinRemoteTrace`. I think it's a good idea, I'll do that depending on how the conversation about keeping the recorder in the `Context` goes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-KbfXVYetM1UKmbd6lDr:b-ofgf1r)*
<!-- Sent from Reviewable.io -->

knz commented:
I am writing this comment at the point where I have reviewed all commits but the last one.
Three general comments:

0. I like the general direction where this is going. Thank you for tackling this. I appreciate the insane amount of work that went into it.

1. please split off all the commits but the last in a separate PR  "refactor the working of EXPLAIN(TRACE)", and rename this PR to "Replace EXPLAIN(TRACE) by a session-wide SQL tracing facility". I believe that all the stars are aligned to get that first PR chunk merged soon, whereas I sense we're still a few design iteration from merging the latter.

2. as a bit of friendly advice, it would make review much easier if you were a little bit more disciplined about what goes in each commit -- here in particular, I think there are two stray patches in the 1st commit, and the commit messages in general are a bit lightweight in content: I miss context and intent. 

Other than that I am 100% on board with the idea to use snowballl tracing to support EXPLAIN(TRACE) and its upcoming replacement. Your changes are good to go from my side for all the commits except the last (modulo the two patches that may need to migrate from one commit to another). I can't speak for the tracing code changes, it's not my area.

I'll look at the session-based tracing in my next comment.

---

Reviewed 8 of 8 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, 9 of 9 files at r4, 3 of 3 files at r5, 1 of 13 files at r6.
Review status: 15 of 27 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/planner.go, line 180 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbe2wayPFR0FoO7QD6a:-Kbe2wazzJibYTE0-ke1:buk458u) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/sql/planner.go#L180)):*
> ```Go
> }
> 
> // hijackCtx changes the current transaction's context to the provided one and
> ```

This specific change does not belong to the 1st commit of the PR, I think.

---

*[pkg/sql/session.go, line 420 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbe3-tnoW6dI79XP8fq:-Kbe3-tnoW6dI79XP8fr:b-n3asfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/sql/session.go#L420)):*
> ```Go
> }
> 
> // hijackCtx changes the transaction's context to the provided one and returns a
> ```

This specific change does not belong to the 1st commit of the PR either, I think.

---

*[pkg/sql/trace_test.go, line 93 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbgVAo1MGfD0YReblJO:-KbgVAo2lG-0u9UOK-gD:bob4vwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7b62c926b2a18f7645bf35e646d85e183481a0/pkg/sql/trace_test.go#L93)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	rows, err := sqlDB.Query(`EXPLAIN (TRACE) SELECT * FROM test.foo`)
> ```

Why did you replace the INSERT test? Seems more interesting to me.

---

*[pkg/util/tracing/tracer.go, line 15 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbgVLR_Fh9VzqqQEAmx:-KbgVLR_Fh9VzqqQEAmy:b-t1auvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7b62c926b2a18f7645bf35e646d85e183481a0/pkg/util/tracing/tracer.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> //
> // Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
> ```

Perhaps use the CRL address for both now. If I'm not mistaken Toby used his other address because he wrote this before he was employed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-KbgVz_Qa5rT-uKAK7H5:b-jpomx1)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 12 of 13 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/session_info.go, line 33 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbgYeVOFqhNqP-vLnNU:-KbgYeVPfUrIkw2qU8lh:b-qgippi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0ffb365d56c2fd51fc07f47b95ffdcd582cd193/pkg/sql/session_info.go#L33)):*
> ```Go
> // sessionInfo is a virtual database exposing per-session information.
> var sessionInfo = virtualSchema{
> 	name: "session_info",
> ```

Yes, yes, but no.
Yes - we need a new virtual schema for CRDB-specific stuff.
Yes - the overall structure of this file is OK.
No - Let's not name it "session_info". This name is *much* too specific. We have "session_info" now, what then? Soon, "transaction_info", "storage_info", and the week after "user_info", before we know it all queries will be slowed down by the time needed to resolve identifiers, because we have so much entries in the namespace table.

So let's do this a bit more forward-thinking and go for a name we'll reuse for the other things.
I propose "crdb_internal" of course since #10317 was already planning to use it. 
The other direction that has dt's and nathan's preference is to create the ability to add virtual tables to the "system" schema, which may be a simple matter of changing the lookup rules in data_source.go to check the virtual schema called "system" in addition to the physical descriptor. May need more thinking.

Finally, just saying, but if you were a little less susceptible to NIH you'd have cherry-picked all the test changes due to a new schema name from #10317 and not have to deal with them yourself. Oh well, whatever suits your boat :angel: 

---

*[pkg/sql/session_info.go, line 43 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbg_4rbarIUBqQZnYDm:-Kbg_4rd2IxpLM8ZuFUt:bmkbvc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0ffb365d56c2fd51fc07f47b95ffdcd582cd193/pkg/sql/session_info.go#L43)):*
> ```Go
> var sessionInfoLatestTraceTable = virtualSchemaTable{
> 	schema: `
> CREATE TABLE session_info.latest_trace (
> ```

Table name: "session_trace".

Whether the table will only hold the latest trace or whether we can let it accumulate multiple traces over time is a matter of implementation. No need to give implementation details in the name.

---

*[pkg/sql/set.go, line 35 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbg_MieJ1_Y1mN9LVU5:-Kbg_MigmGB4lseH1vdo:bgoz9wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0ffb365d56c2fd51fc07f47b95ffdcd582cd193/pkg/sql/set.go#L35)):*
> ```Go
> // Privileges: None.
> //   Notes: postgres/mysql do not require privileges for session variables (some exceptions).
> func (p *planner) Set(n *parser.Set) (planNode, error) {
> ```

This method is now big enough that it ought to be split in multiple setters, one per option, in a suitably defined map[string]option. Please do so.

---

*[pkg/sql/set.go, line 125 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbg_eljK5PMoWMw2NhQ:-Kbg_elllCCZrQ9z8X02:by0zvzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0ffb365d56c2fd51fc07f47b95ffdcd582cd193/pkg/sql/set.go#L125)):*
> ```Go
> 		case parser.ReNormalizeName("on"):
> 			if p.session.TxnState.State != Open || !p.session.TxnState.implicitTxn {
> 				// It'd be nice in principle to be able to start a trace while in a
> ```

Please start this comment by explaining what happens in the normal case.

Also copy this entire comment to the commit message, and extend the commit message around it to explain the new feature: what it does, what it replaces, why it's useful, how to use it, what's possible and what's not possible, etc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221)*
<!-- Sent from Reviewable.io -->

knz commented:
Regarding the SET-based approach in the last commit.

I can't stress enough how important that you expand the commit message quite a bit before you can get a proper review on that change. Without an extensive definition of what the patch is intended to achieve it is hard to look at the code and check whether it does what it should (right now it's hard to read between the lines and guess what it really should do). The comment pertaining to the impossibility of enabling tracing in the middle of transaction is hinting at a constraint, but I'd like to read for example what you want to happen if you issue the SET in a retryable autocommit statement packet, and then see if your code validates what you want it to do.

andreimatei commented:
@knz, I took the `SET TRACE=ON` commit away as you suggested; I'll send it out separately. I'll address all your comments there.

Current commits are r7-r10.

---

Review status: 3 of 17 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/server/node.go, line 898 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbfRtzKHRsAueXY_dnH:-Kbhk6J4_6lKcUuSvQ7l:b8nf1hv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/server/node.go#L898)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

correct. `JoinRemoteTrace` is the only case where a Recorder needs to be used here (there could otherwise be a Recorder active, e.g. in the case of a local "rpc call", but this function would not be responsible for doing anything special with it in that case).
Yes, it could could be returned by `JoinRemoteTrace`. I think it's a good idea, I'll do that depending on how the conversation about keeping the recorder in the `Context` goes.
</blockquote></details>

made the change to stop reading the recorder from the context here.

---

*[pkg/sql/planner.go, line 180 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbe2wayPFR0FoO7QD6a:-KbhLONU5ofHVsYL7rfq:bxmj1od) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/sql/planner.go#L180)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This specific change does not belong to the 1st commit of the PR, I think.
</blockquote></details>

yes, moved to "sql: rework how EXPLAIN(TRACE) works"
Rebasing all these commits was a pain because they were older and stuff changed. I messed up during this process.

---

*[pkg/sql/session.go, line 420 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbe3-tnoW6dI79XP8fq:-KbhLlX0mSi071ZCT42W:b-7mmp82) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacdec02d696876b837c3ec2af1b491643bc941/pkg/sql/session.go#L420)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This specific change does not belong to the 1st commit of the PR either, I think.
</blockquote></details>

moved

---

*[pkg/sql/trace_test.go, line 93 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbgVAo1MGfD0YReblJO:-KbhM0iuVYo0onHUEQIg:b-58e0k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7b62c926b2a18f7645bf35e646d85e183481a0/pkg/sql/trace_test.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why did you replace the INSERT test? Seems more interesting to me.
</blockquote></details>

the INSERT test was completely broken. No data was inserted and the spans collected were irrelevant and misleading. See #9564 and #9563. To paraphrase the streets around US now - "hey / ho / EXPLAIN(DEBUG) has got to go".

But I do really agree that some interesting tests around tracing would be nice. I have this neat util in the works with which you can assert that various things happened or didn't happen to your statements - e.g. all the tests that do transaction restarts can use it to check that the transactions were actually restarted. It was momentarily pushed back because of disagreements about how tests that depend on running multiple statements on a single session should go about it (libpq vs pgx).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-KbhjWYQ_dw6VAnvYTkB:b-y1w98b)*
<!-- Sent from Reviewable.io -->

tamird commented:
1st commit message:

> - introduce RecorderTrace - a canonical "recorder callback" to pass to
>   basictracers. Besides clients not having to writer recorders, the idea
>   is also to provide a good location for the recorded spans to be
>   accumulated, in lieu of each client choosing a more or less arbitrary
>   location to store them.
> - make TeeTracer more flexible - allow a TeeSpan to own or not own each
>   of the children spans. Not owning will the useful on the client-side,
>   next commit.

The grammar here is broken;  "not having to writer recorders" and "useful on the client-side, next commit".

Did not review the SQL changes in depth. The tracing changes seems to complicate things, but I don't have concrete suggestions.

---

Reviewed 3 of 8 files at r1, 24 of 24 files at r7, 3 of 3 files at r8.
Review status: 4 of 17 files reviewed at latest revision, 32 unresolved discussions, some commit checks failed.

---

*[pkg/kv/transport.go, line 169 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjwj5nr_Pxb4isuUiv:-Kbjwj5nr_Pxb4isuUiw:b-np51wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/kv/transport.go#L169)):*
> ```Go
> 	// Fork the original context as this async send may outlast the
> 	// caller's context.
> 	// TODO(andrei): It's sketchy that these spans can outlast the caller's
> ```

are you saying that the span shouldn't be forked? I'd prefer if you spelled it out here.

---

*[pkg/server/node.go, line 801 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjyLMCAeX6oqThJz05:-KbjyLMCAeX6oqThJz06:b-kxjfia) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L801)):*
> ```Go
> 
> 	if err := n.stopper.RunTaskWithErr(func() error {
> 		opCtx, finishSpan := n.setupSpanForIncomingRPC(ctx, args.TraceContext)
> ```

i think you can call it `ctx` here, and it's better to shadow in this case to avoid accidentally using the wrong context in this scope.

---

*[pkg/server/node.go, line 852 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx1hNUu2PeyQWX_2g:-Kbjx1hNUu2PeyQWX_2h:b-7c31kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L852)):*
> ```Go
> 
> // setupSpanForIncomingRPC takes a context and returns a derived context with a
> // new span in it. Depending on input context, that span might be a root span or
> ```

missing article "the"?

---

*[pkg/server/node.go, line 856 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx7MPSWNatncr9xXS:-Kbjx7MPSWNatncr9xXT:b-oz5lrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L856)):*
> ```Go
> // remote span. Note that supporting both the "child of local span" and "child
> // of remote span" cases are important, as this RPC can be called either through
> // the network of directly if the caller is local.
> ```

s/of/or/

---

*[pkg/server/node.go, line 858 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjxA5CItEO9Yno9Hrx:-KbjxA5CItEO9Yno9Hry:bd3joo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L858)):*
> ```Go
> // the network of directly if the caller is local.
> //
> // remoteTranceContext represents the span context of this remote call. Can be
> ```

s/represents/is/

---

*[pkg/server/node.go, line 861 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjxMdXc-A2nidAyqbs:-KbjxMdXc-A2nidAyqbt:b-esv3zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L861)):*
> ```Go
> // nil if this call is not remote.
> //
> // It returns the derived context and a cleanup function to be called when
> ```

this paragraph partially repeats the first paragraph =/

---

*[pkg/server/node.go, line 868 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjxSl_vWCqAbf889Hc:-KbjxSl_vWCqAbf889Hd:b-shssr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L868)):*
> ```Go
> func (n *Node) setupSpanForIncomingRPC(
> 	ctx context.Context, remoteTraceContext *tracing.SpanContextCarrier,
> ) (context.Context, func(br *roachpb.BatchResponse)) {
> ```

s/br //

---

*[pkg/server/node.go, line 871 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx_jzHl4HeV740sU3:-Kbjx_jzHl4HeV740sU4:bd7uw6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L871)):*
> ```Go
> 	const opName = "node.Batch"
> 	tr := n.storeCfg.AmbientCtx.Tracer
> 	opCtx := n.storeCfg.AmbientCtx.AnnotateCtx(ctx)
> ```

is this appropriate? i guess this method has just one caller, but i think this line belongs in the caller.

---

*[pkg/server/node.go, line 904 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjy7M4EddeTXzc599P:-Kbjy7M4EddeTXzc599Q:bhet1lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L904)):*
> ```Go
> 					log.Warning(ctx, err)
> 				}
> 				br.CollectedSpans = append(br.CollectedSpans, encSp)
> ```

probably don't want to do this when `err` above is non-nil.

---

*[pkg/util/tracing/tee_tracer.go, line 77 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk18YjrFEHROMEnL7N:-Kbk18YjrFEHROMEnL7O:bsw0pxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tee_tracer.go#L77)):*
> ```Go
> 		spans[i] = spanWithOpt{
> 			Span:  t.tracers[i].StartSpan(operationName, &o),
> 			Owned: true}
> ```

brace on next line, please.

---

*[pkg/util/tracing/tee_tracer.go, line 94 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk1AysA8jYSmTGhwkV:-Kbk1AysA8jYSmTGhwkW:bc8402q) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tee_tracer.go#L94)):*
> ```Go
> func (t *TeeTracer) CreateSpanFrom(spans ...spanWithOpt) *TeeSpan {
> 	if len(spans) != len(t.tracers) {
> 		panic("non-matching spans for TeeTracer")
> ```

log something useful? lengths, even?

---

*[pkg/util/tracing/tracer.go, line 15 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbgVLR_Fh9VzqqQEAmx:-KbjyTfo7WzMnwS-OX-N:bn7vsyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7b62c926b2a18f7645bf35e646d85e183481a0/pkg/util/tracing/tracer.go#L15)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Perhaps use the CRL address for both now. If I'm not mistaken Toby used his other address because he wrote this before he was employed.
</blockquote></details>

No, Tobi continues to use his gmail, as do I.

---

*[pkg/util/tracing/tracer.go, line 400 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk-Z8r8arQzjKSHmIL:-Kbk-Z8r8arQzjKSHmIM:b-szwe4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L400)):*
> ```Go
> // trace.
> func NewRecordingTracer() (opentracing.Tracer, *RecordedTrace) {
> 	rec := RecordedTrace{}
> ```

nit: `rec := &RecordedTrace{}` or `rec := new(RecordedTrace)`

---

*[pkg/util/tracing/tracer.go, line 411 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk-gjqz4CBv90yWRBF:-Kbk-gjqz4CBv90yWRBG:b-9icfqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L411)):*
> ```Go
> }
> 
> // createTeeSpanWithRecorder tee-s a current span. It creates a per-request
> ```

s/tee-s/tees/

---

*[pkg/util/tracing/tracer.go, line 417 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk-wKyYvpqszQw-ZqO:-Kbk-wKyYvpqszQw-ZqP:b-o40eex) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L417)):*
The first paragraph states

> The caller takes ownership and is required to Finish() it.

Probably not a good idea to bury the subtleties of the interface in a second paragraph, particularly, when they entirely contradict the first paragraph.

---

*[pkg/util/tracing/tracer.go, line 432 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0NFHxg84oopEOic5:-Kbk0NFHxg84oopEOic6:bw8z65m) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L432)):*
> ```Go
> 	// The span's tracer needs to be the first tracer in the Tee, and the
> 	// corresponding span needs to be the first argument to CreateSpanFrom, since
> 	// we want the span's SpanContext to be serialized when RPCs will be made.
> ```

s/will be/have been/

---

*[pkg/util/tracing/tracer.go, line 437 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0C8uYwo28tVmU0El:-Kbk0C8uYwo28tVmU0Em:bg6thsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L437)):*
> ```Go
> 		spanWithOpt{
> 			Span:  span,
> 			Owned: adopt},
> ```

closing brace on the next line, please.

---

*[pkg/util/tracing/tracer.go, line 440 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0Dqyd6pAA0-PIyV-:-Kbk0Dqyd6pAA0-PIyV0:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L440)):*
> ```Go
> 		spanWithOpt{
> 			Span:  recordingSpan,
> 			Owned: true})
> ```

ditto

---

*[pkg/util/tracing/tracer.go, line 480 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0dxudn95EL4CG7XQ:-Kbk0dxudn95EL4CG7XR:bg7eor0) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L480)):*
> ```Go
> 	switch err {
> 	case nil:
> 		// TODO(andrei): should we use ChildOf instead of FollowsFrom?
> ```

how can we answer this?

---

*[pkg/util/tracing/tracer.go, line 483 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0h1rKdIDFyvFk8Ex:-Kbk0h1rKdIDFyvFk8Ey:bzhkz4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L483)):*
> ```Go
> 		sp = tr.StartSpan(opName, opentracing.FollowsFrom(wireContext))
> 		// Copy baggage items to tags so they show up in the Lightstep UI.
> 		sp.Context().ForeachBaggageItem(func(k, v string) bool { sp.SetTag(k, v); return true })
> ```

use a few more lines; they're cheap.

---

*[pkg/util/tracing/tracer.go, line 485 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjysEENR7wmi9rv2Zj:-KbjysEENR7wmi9rv2Zk:b-m5tch4) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L485)):*
> ```Go
> 		sp.Context().ForeachBaggageItem(func(k, v string) bool { sp.SetTag(k, v); return true })
> 		sp.LogFields(otlog.String("event", opName))
> 	case opentracing.ErrSpanContextNotFound:
> ```

what is the point of this `case`?

---

*[pkg/util/tracing/tracer.go, line 490 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0pGlJy2KcoBTwU_B:-Kbk0pGlJy2KcoBTwU_C:b-p1brrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L490)):*
> ```Go
> 		return ctx, nil, err
> 	}
> 	if sp.BaggageItem(Snowball) == "" {
> ```

nit: check for `== "1"`?

---

*[pkg/util/tracing/tracer.go, line 500 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk12WVThQQh1GlGHUb:-Kbk12WVThQQh1GlGHUc:b-nu3a8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L500)):*
> ```Go
> 		return ctx, nil, errors.Errorf("there's already a recorder in the ctx")
> 	}
> 	teeSpan, recorder := createTeeSpanWithRecorder(sp, true /* adopt */, opName)
> ```

is adopt ever false?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-KbjwHrk5bEa4rCBZelB:ba7y38i)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Cool stuff! Did you test that snowball traces now also show up on lightstep?

---

Review status: 4 of 17 files reviewed at latest revision, 32 unresolved discussions, some commit checks failed.

---

*[pkg/util/tracing/tee_tracer.go, line 86 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbolDJpHp9Cq56tHQdW:-KbolDJpHp9Cq56tHQdX:b9b44fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tee_tracer.go#L86)):*
> ```Go
> 	// If Owned is set, the span needs to be Finish()ed by the TeeSpan that
> 	// contains it.
> 	Owned bool
> ```

any reason this is exported?

---

*[pkg/util/tracing/tracer.go, line 384 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbojCypdQxlgaYg9e85:-KbojCypdQxlgaYg9e86:b-vby5qq) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L384)):*
> ```Go
> // WithRecorder returns a context derived from the given context, for which
> // RecorderFromCtx returns the given recorder.
> func WithRecorder(ctx context.Context, rec *RecordedTrace) context.Context {
> ```

We can unexport `WithRecorder` and `RecorderFromCtx` now.

---

*[pkg/util/tracing/tracer.go, line 480 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0dxudn95EL4CG7XQ:-KbohHw7SzJgb283OsOm:b-nn2j20) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L480)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how can we answer this?
</blockquote></details>

AFAICT ChildOf is intended to be used with spans that the parent operation depends on (and thus will be finished before the parent span); FollowsFrom with spans that might outlive the parent. I couldn't find any actual difference in handling in the actual implementations though (basictracer/lightstep).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-Kbon2BEbrkk0r62gcOd:b-8lsuh3)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 17 files reviewed at latest revision, 32 unresolved discussions, some commit checks pending.

---

*[pkg/kv/transport.go, line 169 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjwj5nr_Pxb4isuUiv:-Kc4JT-6iu_n4Zn_AyQA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/kv/transport.go#L169)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

are you saying that the span shouldn't be forked? I'd prefer if you spelled it out here.
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 801 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjyLMCAeX6oqThJz05:-Kc4JtXEjjVeor962EvV:by2y2hr) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L801)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think you can call it `ctx` here, and it's better to shadow in this case to avoid accidentally using the wrong context in this scope.
</blockquote></details>

But the linter yells at me...

---

*[pkg/server/node.go, line 852 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx1hNUu2PeyQWX_2g:-Kc4KoGwgtCb4UpVreEJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L852)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

missing article "the"?
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 856 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx7MPSWNatncr9xXS:-Kc4KrrFZ04CB-fKjugu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L856)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/of/or/
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 858 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjxA5CItEO9Yno9Hrx:-Kc4L2qz42Gh71-Gilgo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L858)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/represents/is/
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 861 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjxMdXc-A2nidAyqbs:-Kc4L6CXhyfGEdIL215x:b-t82dda) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L861)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this paragraph partially repeats the first paragraph =/
</blockquote></details>

I think it's fine... what would you change?

---

*[pkg/server/node.go, line 868 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjxSl_vWCqAbf889Hc:-Kc4L8d4hJgP_fypCEdq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L868)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/br //
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 871 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx_jzHl4HeV740sU3:-Kc4MheqnQB11KCA4Vx7:b4fw9i6) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L871)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this appropriate? i guess this method has just one caller, but i think this line belongs in the caller.
</blockquote></details>

I think this was just redundant; the caller already annotates the ctx

---

*[pkg/server/node.go, line 904 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjy7M4EddeTXzc599P:-Kc4Mvzi6LTl5vw66JMm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L904)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

probably don't want to do this when `err` above is non-nil.
</blockquote></details>

Done.

---

*[pkg/util/tracing/tee_tracer.go, line 77 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk18YjrFEHROMEnL7N:-Kc4NF4Ulf3TueJABr1p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tee_tracer.go#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

brace on next line, please.
</blockquote></details>

Done.

---

*[pkg/util/tracing/tee_tracer.go, line 86 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbolDJpHp9Cq56tHQdW:-Kc4NNqy5ymmG2_jVIVL:b-6fwj30) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tee_tracer.go#L86)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

any reason this is exported?
</blockquote></details>

no reason. changed

---

*[pkg/util/tracing/tee_tracer.go, line 94 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk1AysA8jYSmTGhwkV:-Kc4N_uX9wt11wszcmHO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tee_tracer.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

log something useful? lengths, even?
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 384 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbojCypdQxlgaYg9e85:-Kc4OOzcXWZul5KasDwD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L384)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can unexport `WithRecorder` and `RecorderFromCtx` now.
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 400 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk-Z8r8arQzjKSHmIL:-Kc4Oc5DK3pkp8pFGM3f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L400)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `rec := &RecordedTrace{}` or `rec := new(RecordedTrace)`
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 411 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk-gjqz4CBv90yWRBF:-Kc4OmFsQlBJMGw_YfA9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L411)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/tee-s/tees/
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 417 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk-wKyYvpqszQw-ZqO:-Kc4OwclY-ckaTA6fH_J:b-4f2c4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L417)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The first paragraph states

> The caller takes ownership and is required to Finish() it.

Probably not a good idea to bury the subtleties of the interface in a second paragraph, particularly, when they entirely contradict the first paragraph.
</blockquote></details>

but there's no contradiction. The caller always takes ownership of the returned span. The question is whether the ownership of the passed-in span changes.
I've changed "the caller is no longer in charge of Finish()ing that span" to "...  the passed-in span". Does that help?

---

*[pkg/util/tracing/tracer.go, line 432 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0NFHxg84oopEOic5:-Kc4PYKIv60LD7Nl9BZh:bdm8bcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L432)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/will be/have been/
</blockquote></details>

I changed to "are invoked". It's about the future, not the past.

---

*[pkg/util/tracing/tracer.go, line 440 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0Dqyd6pAA0-PIyV-:-Kc4PeZvIGtAgjMsS8fH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L440)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 480 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0dxudn95EL4CG7XQ:-Kc4PnUckumQCxYE53rm:bkywhx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L480)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

AFAICT ChildOf is intended to be used with spans that the parent operation depends on (and thus will be finished before the parent span); FollowsFrom with spans that might outlive the parent. I couldn't find any actual difference in handling in the actual implementations though (basictracer/lightstep).
</blockquote></details>

Right. So which one do you think is appropriate here? The (forked) span in the kv transport does depend on this. But, being a remote call, might not wait for it...

---

*[pkg/util/tracing/tracer.go, line 483 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0h1rKdIDFyvFk8Ex:-Kc4QVUAxWPt-j1hjHbO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L483)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

use a few more lines; they're cheap.
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 485 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjysEENR7wmi9rv2Zj:-Kc4QXoKYWcspNs9GUTy:b-yjnugv) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L485)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what is the point of this `case`?
</blockquote></details>

to make the reader aware of this possible error, and of the fact that it's unexpected for us for this function to have been called when carrier doesn't contain a span.

---

*[pkg/util/tracing/tracer.go, line 490 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0pGlJy2KcoBTwU_B:-Kc4ROpU-AhfWfUkDuw1:b4e6pdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L490)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: check for `== "1"`?
</blockquote></details>

well the definition of the `Snowball` constant says that it is "set" on a snowball trace. I'd rather keep that and treat it as present/absent.

---

*[pkg/util/tracing/tracer.go, line 500 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk12WVThQQh1GlGHUb:-Kc4RioD_7qRbi0c6aeu:b-299y1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L500)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is adopt ever false?
</blockquote></details>

yes, in the next commits

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 9 files at r9, 6 of 11 files at r11, 2 of 3 files at r12.
Review status: 12 of 17 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/server/node.go, line 801 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjyLMCAeX6oqThJz05:-Kc5PLpAww3Q6llo98gX:b-5l7er9) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L801)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

But the linter yells at me...
</blockquote></details>

yeah, you'll have to `var finishSpan`.

---

*[pkg/server/node.go, line 871 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx_jzHl4HeV740sU3:-Kc5PYBz3TCApNmL48rd:b6dsaj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L871)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this was just redundant; the caller already annotates the ctx
</blockquote></details>

what's the point of this alias?

---

*[pkg/util/tracing/tracer.go, line 437 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0C8uYwo28tVmU0El:-Kc5Q7MqlXVdrqNqp_LN:b-a2jkrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L437)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

closing brace on the next line, please.
</blockquote></details>

missed this one

---

*[pkg/util/tracing/tracer.go, line 483 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0h1rKdIDFyvFk8Ex:-Kc5QDygQTZzOGm8pfkt:by4ch68) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L483)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`sp.Context().ForeachBaggageItem(func(k, v string) bool {`
to match that the closing brace and paren are on the same line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
@radu yup verified on lightstep. `EXPLAIN(TRACE)` traces are now there.

---

Review status: 2 of 17 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/server/node.go, line 801 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-KbjyLMCAeX6oqThJz05:-Kc6GInlIcY2E-8KCiHF:b6szshj) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L801)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

yeah, you'll have to `var finishSpan`.
</blockquote></details>

so like this?

---

*[pkg/server/node.go, line 871 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbjx_jzHl4HeV740sU3:-Kc6HIkrZL8WG6VtWgNy:b2yh1k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/server/node.go#L871)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's the point of this alias?
</blockquote></details>

not sure what the thinking was but doesn't appear to be necessary. removed, thanks.

---

*[pkg/util/tracing/tracer.go, line 483 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0h1rKdIDFyvFk8Ex:-Kc6Hdcjl5MEzX5-DrCx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L483)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`sp.Context().ForeachBaggageItem(func(k, v string) bool {`
to match that the closing brace and paren are on the same line
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-Kc6JoOI6C4FLay4aKKp:b27uid3)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
TFTR y'all

---

Review status: 2 of 17 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-Kc6JyHoO0nAVfHzQ9R5:b-pttas8)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice!

---

Review status: 2 of 17 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/util/tracing/tracer.go, line 480 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0dxudn95EL4CG7XQ:-KcDb0GDkLviWnPs3VlO:buuxqsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L480)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Right. So which one do you think is appropriate here? The (forked) span in the kv transport does depend on this. But, being a remote call, might not wait for it...
</blockquote></details>

I think FollowsFrom is better given that we're not putting any constraints on what the caller does. I don't feel strongly though, it doesn't really matter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-:-KcDbmOvaAff7M2vlN1I:b-83sjoh)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 17 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/util/tracing/tracer.go, line 480 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221#-Kbk0dxudn95EL4CG7XQ:-KcK2AvYtdPuRq6DLSXB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555f3a421a264c3fa8ae049c0cb1591d474814d2/pkg/util/tracing/tracer.go#L480)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think FollowsFrom is better given that we're not putting any constraints on what the caller does. I don't feel strongly though, it doesn't really matter.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13221)*
<!-- Sent from Reviewable.io -->
